### DISCUSSION TITLE
Api to get all routes

### DISCUSSION THREAD

**Discussion by EralChen** [vikejs/vike#1411] [Discussion] [answered]
hello. Is there an api  to get  all routes ? I saw it in `pageContext._pageRoutes`  .

![image](https://github.com/vikejs/vike/assets/82485978/2e6b8ec0-9ba1-4c49-9c25-ea329b851d26)

I would like to gen a menu using _pageRoutes .

But it might suffer from a breaking change some day. 

Is there a more suitable way to do that? 

Additionally，Can I add some metadata for a route like vue-route's usage ？

![image](https://github.com/vikejs/vike/assets/82485978/49e4ab89-0954-4b69-ac27-697b2d4c2353)

**Comment by nhattran998** [comment, marked as answer]
I think you should create an API route, or use Telefunc RPC to read your `pages` structure --> Transform that like you want.
If you always need that menu, inject it in the `renderPage` function if you using a server. Vice versa, inject it in return in `onRenderHtml` like 

```ts
const onRenderHtml: OnRenderHtmlAsync = async (pageContext) => {
return {
    documentHtml,
    pageContext: {
      menu,
    },
  };
}
```

**Comment by brillout** [comment]
See https://vike.dev/markdown#page-list.
